src/core/mac/mac.cpp

```diff
@@ -2257,39 +2261,41 @@ uint8_t Mac::GetTimeIeOffset(const Frame &aFrame)
 #endif
 
 #if OPENTHREAD_CONFIG_MAC_CSL_RECEIVER_ENABLE
-void Mac::SetCslChannel(uint8_t aChannel)
+void Mac::UpdateCsl(void)
 {
-    VerifyOrExit(GetCslChannel() != aChannel);
+    uint16_t period;
+    uint8_t  channel;
 
-    mLinks.GetSubMac().SetCslChannel(aChannel);
-    mLinks.GetSubMac().SetCslChannelSpecified(aChannel != 0);
+    VerifyOrExit(IsCslSupported());
 
-    if (IsCslEnabled())
+    period  = Get<Mle::Mle>().IsRxOnWhenIdle() ? 0 : GetCslPeriod();
+    channel = GetCslChannel() ? GetCslChannel() : mRadioChannel;
```

Comment:
Another thing to consider:
- Say `GetCslChannel()` is zero (which means we need to use pan/radio channel for CSL)
- And then pan/radio channel get changed.
- We need to somehow trigger this method so to update the channel on `SubMac`

If I understand the existing/previous code, we passed the channel in `CslSample()` as a parameter. Then `SubMac` was checking if the CSL channel was not set before, it used the passed in channel.

Reply:
Good catch! Made the proper change in `Mac::SetPanChannel()`.